flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/assigners/WindowStagger.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.api.windowing.assigners;
+
+import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
+
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * A {@code WindowStagger} staggers offset in runtime for each window assignment.
+ */
+public enum WindowStagger {
+	/**
+	 * Default mode,  all panes fire at the same time across all partitions.
+	 */
+	ALIGNED {
+		@Override
+		public long getStaggerOffset(
+			final long currentProcessingTime,
+			final long size) {
+			return 0L;
+		}
+	},
+
+	/**
+	 * Stagger offset is sampled from uniform distribution U(0, WindowSize) when first event ingested in the partitioned operator.
+	 */
+	RANDOM {
+		@Override
+		public long getStaggerOffset(
+			final long currentProcessingTime,
+			final long size) {
+			return (long) (ThreadLocalRandom.current().nextDouble() * size);
+		}
+	},
+
+	/**
+	 * Stagger offset is the ingestion delay in processing time, which is the difference between first event ingestion time and its corresponding processing window start time
+	 * in the partitioned operator. In other words, each partitioned window starts when its first pane created.
```

Comment:
I think this is incorrect and should be more like the description of `RANDOM`. We determine the stagger offset when "the first event is ingested in the partitioned operator". It's not determined for each pane.

Reply:
Reworded